src/easylink/step.py

```diff
@@ -18,135 +31,262 @@ class Step(ABC):
     """Basic Step for result node."""
 
-    def __init__(self, name: str, **params) -> None:
-        self.name = name
-        self.input_validator = params["input_validator"]
-        self.out_dir = params["out_dir"]
+    def get_implementation_graph(
+        self, graph, pipeline_config: LayeredConfigTree
+    ) -> nx.MultiDiGraph:
+        graph.add_node("results")
 
-    def get_subgraph(self, config: "Config") -> MultiDiGraph:
-        sub_graph = MultiDiGraph()
-        sub_graph.add_node(
-            "results", input_validator=self.input_validator, out_dir=self.out_dir
-        )
-        return sub_graph
+    def update_edges(self, graph, pipeline_config: LayeredConfigTree) -> None:
+        for _, sink, edge_attrs in graph.out_edges(self.name, data=True):
+            graph.add_edge(
+                "results",
+                sink,
+                input_slot=edge_attrs["input_slot"],
+                output_slot=edge_attrs["output_slot"],
+            )
+
+        for source, _, edge_attrs in graph.in_edges(self.name, data=True):
+            graph.add_edge(
+                source,
+                "results",
+                input_slot=edge_attrs["input_slot"],
+                output_slot=edge_attrs["output_slot"],
+            )
+
+    def validate_step(self, pipeline_config: LayeredConfigTree) -> Dict[str, List[str]]:
+        return {}
 
 
 class ImplementedStep(Step):
     """Step for leaf node tied to a specific single implementation"""
 
-    def __init__(self, name: str, **params) -> None:
-        self.name = name
-        self.input_validator = params["input_validator"]
-        self.out_dir = params["out_dir"]
-
-    def get_subgraph(self, config: "Config") -> MultiDiGraph:
+    def get_implementation_graph(
+        self, graph, pipeline_config: LayeredConfigTree
+    ) -> Implementation:
         """Return a single node with an implementation attribute."""
-        sub = MultiDiGraph()
-        implementation_name = config.get_implementation_name(self.name)
-        implementation_config = config.pipeline[self.name]["implementation"]["configuration"]
+        implementation_name = pipeline_config[self.name]["implementation"]["name"]
+        implementation_config = pipeline_config[self.name]["implementation"]["configuration"]
         implementation = Implementation(
             name=implementation_name,
             step_name=self.name,
             environment_variables=implementation_config.to_dict(),
         )
-        sub.add_node(
+        graph.add_node(
             implementation_name,
             implementation=implementation,
-            input_validator=self.input_validator,
-            out_dir=self.out_dir / implementation_name,
+            # out_dir=Path("intermediate") / implementation_name,
         )
-        return sub
+
+    def update_edges(self, graph, pipeline_config: LayeredConfigTree) -> None:
+        implementation_name = pipeline_config[self.name]["implementation"]["name"]
```

Comment:
Is this depending on global uniqueness of implementation name?

Reply:
I think it's only depending on uniqueness at the same level of hierarchy, since "pipeline_config" in this case is generally a sub-config. maybe calling it "step_config" would make it clearer that this is the case.

Reply:
But isn't `graph` the entire graph?

I don't understand how calling `graph.add_edge(implementation_name, ...)` below could work if implementation_name is not unique.

Reply:
Ah! Yes, you are right, sorry to be a bit slow on the uptake there.